README.md
AUTHORS:
--------
Hector Garcia Huerta hectorgh@gmail.com  
Raul Rodrigo Segura raurodse@gmail.com  
What do i need for starting?
-------------------------
* Make file key in folder /etc/n4d/ containing a master secret key. This file is only readable and writable by root user. On Linux you can execute :

        cat /dev/urandom| tr -dc '0-9a-zA-Z' |{ head -c 50;echo ""; } > /etc/n4d/key  
        chmod 400 /etc/n4d/key  
        chown root:root /etc/n4d/key

* You need create digital certificate, because n4d uses ssl communication. You must create key which you will sign certificates and a certificate. These move to folder /etc/n4d/cert.
Previously you has created folder /etc/n4d/cert. You can use openssl for create this digital certificate.

        mkdir /etc/n4d/cert/
        openssl genrsa -out /etc/n4d/cert/n4dkey.pem 2048
        yes '' | openssl req -new -key /etc/n4d/cert/n4dkey.pem -out /etc/n4d/cert/n4d.csr
        openssl x509 -req -days 600 -in /etc/n4d/cert/n4d.csr -signkey /etc/n4d/cert/n4dkey.pem -out /etc/n4d/cert/n4dcert.pem

For installing
--------------

	./setup.py install --install-data=/




debian/changelog
n4d (0.81.4) precise; urgency=low

  * UNRELEASED

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Tue, 30 Jul 2013 13:09:43 +0200

n4d (0.81.3) precise; urgency=high

  * Fixed offline_getvalues

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Thu, 25 Jul 2013 13:21:46 +0200

n4d (0.81.2) precise; urgency=high

  * First release without server, client and lliurex plugins
  * Fixed offline_getvalues in n4d-vars
  * n4d-vars uses n4d service when it is running, otherwise it uses offline_getvalues

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Tue, 23 Jul 2013 11:15:17 +0200

n4d (0.81) precise; urgency=low

  * Basic print changes on n4d-server
  * Fixed n4d-create-conf conf output path
  * n4d-vars reads variables even with n4d stopped

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Mon, 08 Jul 2013 09:33:51 +0200

n4d (0.80.9) precise; urgency=low

  * Now it's really fixed

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Mon, 08 Jul 2013 09:22:42 +0200

n4d (0.80.8) precise; urgency=low

  * Fixed password change

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Thu, 04 Jul 2013 18:35:06 +0200

n4d (0.80.7) precise; urgency=high

  * Fixed teachers groups. No longer tries to add to tcos groups

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Tue, 02 Jul 2013 13:18:19 +0200

n4d (0.80.6) precise; urgency=high

  * Fixed USER NOT FOUND

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Mon, 01 Jul 2013 08:11:40 +0200

n4d (0.80.5) precise; urgency=low

  * Fixed harvester functionality

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Mon, 01 Jul 2013 08:10:43 +0200

n4d (0.80.4) precise; urgency=high

  * Fixed gescen importation

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Wed, 26 Jun 2013 14:41:49 +0200

n4d (0.80.3) precise; urgency=high

  * Fixed golem perms

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Thu, 20 Jun 2013 17:58:25 +0200

n4d (0.80.2) precise; urgency=low

  * Fixed minor bug

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Thu, 20 Jun 2013 17:41:53 +0200

n4d (0.80.1) precise; urgency=low

  * Fixed NetShareOperations with the new /net path
  * Fixed VariablesManager for clients

 -- Hector Garcia Huerta <cless@anubis.lliurex.fs>  Tue, 18 Jun 2013 13:08:41 +0200

n4d (0.80) precise; urgency=low

  [ Raul Rodrigo Segura ]
  * Correctly assigned rc start and stop
  * n4d-createconf regex fixed
  [ Hector Garcia Huerta ]
  * get_variable full info returns wether it's remote or not
  * fixed ip secured functions
  * Freezing and unfreezing groups back to life

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Mon, 17 Jun 2013 11:05:16 +0200

n4d (0.79.8) precise; urgency=low

  * VariablesManager tries three times to get a remote variable in case of failure to connect.

 -- Hector Garcia Huerta <cless@anubis.lliurex.fs>  Thu, 13 Jun 2013 20:36:08 +0200

n4d (0.79.7) precise; urgency=low

  * Update function on variable

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Thu, 13 Jun 2013 20:27:26 +0200

n4d (0.79.6) precise; urgency=low

  * N4D start on 19 rc. Now is before libnss-ldap 

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Wed, 12 Jun 2013 20:26:40 +0200

n4d (0.79.5) precise; urgency=low

  * Updated by Angel

 -- Angel Berlanas Vicente <angel.berlanas@gmail.com>  Tue, 11 Jun 2013 12:09:21 +0200

n4d (0.79.4) precise; urgency=high

  * now n4d launcher emit n4d signal with no-wait 

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Fri, 08 Jun 2013 10:24:13 +0200

n4d (0.79.3) precise; urgency=high

  * N4D start when networking has started

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Fri, 07 Jun 2013 10:24:13 +0200

n4d (0.79.2) precise; urgency=low

  * Added setgid to users folder

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Fri, 07 Jun 2013 10:21:22 +0200

n4d (0.79.1) precise; urgency=low

  * Fixed Variables Manager bug when using add_variable
  * Fixed NetFilesManager

 -- Hector Garcia Huerta <cless@anubis.lliurex.fs>  Fri, 07 Jun 2013 08:50:52 +0200

n4d (0.78.9) precise; urgency=low

  * Fixed upstart
  * Fixed remote variables
  * Fixed ldapmanager to show admins in llum

 -- Hector Garcia Huerta <cless@anubis.lliurex.fs>  Mon, 03 Jun 2013 22:20:45 +0200

n4d (0.78.8) precise; urgency=low

  * Added log

 -- Hector Garcia Huerta <cless@anubis.lliurex.fs>  Fri, 31 May 2013 10:12:52 +0200

n4d (0.78.7) precise; urgency=low

  * Fixed typo
  * Added is-controlled option to n4d-modules. Writes 'TRUE' or 'FALSE'
  * Added is_controlled_startup to n4d as a basic function. Returns True or False

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Tue, 28 May 2013 14:29:12 +0200

n4d (0.78.6) precise; urgency=low

  * Local admin from server can login on llum 

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Tue, 28 May 2013 14:17:25 +0200

n4d (0.78.5) precise; urgency=low

  * Commented localereset line

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Mon, 27 May 2013 20:07:29 +0200

n4d (0.78.4) precise; urgency=low

  * Semi dummy upgrade

 -- Hector Garcia Huerta <cless@anubis.lliurex.fs>  Mon, 27 May 2013 14:22:26 +0200

n4d (0.78.3) precise; urgency=low

  [ Hector Garcia Huerta ]
  * Fixed locale reset

  [ Raul Rodrigo Segura ]
  * Change path to rsync when create net home folders

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Fri, 24 May 2013 11:35:47 +0200

n4d (0.78.2) precise; urgency=low

  * Teachers are members of epoptes group

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Tue, 07 May 2013 10:20:29 +0200

n4d (0.78.1) precise; urgency=low

  * Fixed https webpage. Showing methods in browser should work again

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Thu, 02 May 2013 08:11:22 +0200

n4d (0.78) precise; urgency=low

  [ Hector Garcia Huerta ]
  * Added folder to use support classes inside plugins. INCEPTION

  [ Raul Rodrigo Segura ]
  * Removed register machine. This is moved to Dnsmasq 

 -- Raul Rodrigo Segura <raurodse@gmail.com>  Mon, 22 Apr 2013 13:47:00 +0200

n4d (0.77) precise; urgency=low


install-files/usr/share/n4d/xmlrpc-server/server.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import locale
locale.setlocale(locale.LC_ALL, 'C.UTF-8')

"""
 *******************************************************************************
 * 
 * $Id: SecureDocXMLRPCServer.py 4 2008-06-04 18:44:13Z yingera $
 * $URL: https://xxxxxx/repos/utils/trunk/tools/SVNRPCServer.py $
 *
 * $Date: 2008-06-04 13:44:13 -0500 (Wed, 04 Jun 2008) $
 * $Author: yingera $
 *
 * Authors: Laszlo Nagy, Andrew Yinger
 *
 * Description: Threaded, Documenting SecureDocXMLRPCServer.py - over HTTPS.
 *
 *  requires pyOpenSSL: http://sourceforge.net/project/showfiles.php?group_id=31249
 *   ...and open SSL certs installed.
 *
 * Based on this article: http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/81549
 *
 *******************************************************************************
"""

import SocketServer
import BaseHTTPServer
import SimpleHTTPServer
import SimpleXMLRPCServer
import imp


import fcntl
import time
import random
import base64
import socket, os
from OpenSSL import SSL
from threading import Event, currentThread, Thread, Condition
from thread import start_new_thread as start
from DocXMLRPCServer import DocXMLRPCServer, DocXMLRPCRequestHandler

import xmlrpclib
import threading
threading._DummyThread._Thread__stop = lambda x: 42

#locale.resetlocale()

# static stuff

'''
DEFAULTKEYFILE='/etc/lliurex-secrets/pki/n4d/n4d.key'	# Replace with your PEM formatted key file
DEFAULTCERTFILE='/etc/lliurex-secrets/certs/n4d/n4d'  # Replace with your PEM formatted certificate file
'''

DEFAULTKEYFILE='/etc/n4d/cert/n4dkey.pem'	# Replace with your PEM formatted key file
DEFAULTCERTFILE='/etc/n4d/cert/n4dcert.pem'  # Replace with your PEM formatted certificate file

TRIGGER_BLOCK=False


class SecureDocXMLRpcRequestHandler(DocXMLRPCRequestHandler):
	"""Secure Doc XML-RPC request handler class.
	It it very similar to DocXMLRPCRequestHandler but it uses HTTPS for transporting XML data.
	"""
	def setup(self):
		self.connection = self.request
		self.rfile = socket._fileobject(self.request, "rb", self.rbufsize)
		self.wfile = socket._fileobject(self.request, "wb", self.wbufsize)

	def address_string(self):
		"getting 'FQDN' from host seems to stall on some ip addresses, so... just (quickly!) return raw host address"
		host, port = self.client_address
		#return socket.getfqdn(host)
		return host

	def do_POST(self):
		"""Handles the HTTPS POST request.
		It was copied out from SimpleXMLRPCServer.py and modified to shutdown the socket cleanly.
		"""
		try:
			# get arguments
			data = self.rfile.read(int(self.headers["content-length"]))
			# In previous versions of SimpleXMLRPCServer, _dispatch
			# could be overridden in this class, instead of in
			# SimpleXMLRPCDispatcher. To maintain backwards compatibility,
			# check to see if a subclass implements _dispatch and dispatch
			# using that method if present.
			addr,num=self.client_address
			response = self.server._marshaled_dispatch(data, getattr(self, '_dispatch', None),client_address=addr)
			
			
		except: # This should only happen if the module is buggy
			# internal error, report as HTTP server error
			self.send_response(500)
			self.end_headers()
		else:
			# got a valid XML RPC response
			self.send_response(200)
			self.send_header("Content-type", "text/xml")
			self.send_header("Content-length", str(len(response)))
			self.end_headers()
			self.wfile.write(response)

			# shut down the connection
			self.wfile.flush()
			'''
			global TRIGGER_BLOCK
			while(TRIGGER_BLOCK):
				print "waiting for trigger_block to be unlocked..."
				time.sleep(int(5*random.random()))
			TRIGGER_BLOCK=True
			print "done with response. Sleeping..."
			time.sleep(5)
			print "done sleeping."
			TRIGGER_BLOCK=False
			'''
			self.connection.shutdown() # Modified here!

	def do_GET(self):
		"""Handles the HTTP GET request.

		Interpret all HTTP GET requests as requests for server
		documentation.
		"""
		# Check that the path is legal

		if not self.is_rpc_path_valid():
			self.report_404()
			return
		
		
		#print self.path
		#response = self.server.generate_html_documentation()
		
		header="<html><body>"
		foot="</body></html>"
		
		response=header
		
		for plugin in l.cm.plugins:
			response+="<b>["+plugin.type+":"+plugin.class_name+"]</b><br>"
			for method in plugin.function:
				try:
					#server.register_function(getattr(l.objects[plugin.class_name],method))
					args=getattr(l.objects[plugin.class_name],method).func_code.co_varnames[:getattr(l.objects[plugin.class_name],method).func_code.co_argcount]
					response+="<pre>      " + method+str(args).replace("'self',","")+" , " + str(getattr(l.objects[plugin.class_name],method).__doc__) + "</pre>"
				except:
					pass
		
		response+=foot
		

		
		self.send_response(200)
		self.send_header("Content-type", "text/html")
		self.send_header("Content-length", str(len(response)))
		self.end_headers()
		self.wfile.write(response)

		# shut down the connection
		self.wfile.flush()
		self.connection.shutdown() # Modified here!

	def report_404 (self):
		# Report a 404 error
		self.send_response(404)
		response = 'No such page'
		self.send_header("Content-type", "text/plain")
		self.send_header("Content-length", str(len(response)))
		self.end_headers()
		self.wfile.write(response)
		# shut down the connection
		self.wfile.flush()
		self.connection.shutdown() # Modified here!



class CustomThreadingMixIn:
	"""Mix-in class to handle each request in a new thread."""
	# Decides how threads will act upon termination of the main process
	daemon_threads = True

	def process_request_thread(self, request, client_address):
		"""Same as in BaseServer but as a thread.
		In addition, exception handling is done here.
		"""
		try:
			self.finish_request(request, client_address)
			self.close_request(request)
		except (socket.error, SSL.SysCallError), why:
			print 'socket.error finishing request from "%s"; Error: %s' % (client_address, str(why))
			self.close_request(request)
		except:
			self.handle_error(request, client_address)
			self.close_request(request)

	def process_request(self, request, client_address):
		"""Start a new thread to process the request."""
		t = Thread(target = self.process_request_thread, args = (request, client_address))
		if self.daemon_threads:
			t.setDaemon(1)
		t.start()



class SecureDocXMLRPCServer(CustomThreadingMixIn, DocXMLRPCServer):
	def __init__(self, registerInstance, server_address, keyFile=DEFAULTKEYFILE, certFile=DEFAULTCERTFILE, logRequests=True):
		"""Secure Documenting XML-RPC server.
		It it very similar to DocXMLRPCServer but it uses HTTPS for transporting XML data.
		"""
		DocXMLRPCServer.__init__(self, server_address, SecureDocXMLRpcRequestHandler, logRequests,True)
		self.logRequests = logRequests

		# stuff for doc server
		try: self.set_server_title(registerInstance.title)
		except AttributeError: self.set_server_title('N4D Documentation page')
		try: self.set_server_name(registerInstance.name)
		except AttributeError: self.set_server_name('N4D')
		#for plugin in registerInstace
		if registerInstance.__doc__: self.set_server_documentation(registerInstance.__doc__)
		else: self.set_server_documentation('default documentation')
		
		
		#self.register_introspection_functions()
		
		# init stuff, handle different versions:
		try:
			SimpleXMLRPCServer.SimpleXMLRPCDispatcher.__init__(self,allow_none=True)
		except TypeError:
			# An exception is raised in Python 2.5 as the prototype of the __init__
			# method has changed and now has 3 arguments (self, allow_none, encoding)
			SimpleXMLRPCServer.SimpleXMLRPCDispatcher.__init__(self, allow_none=True, encoding=None)
		SocketServer.BaseServer.__init__(self, server_address, SecureDocXMLRpcRequestHandler)
		self.register_instance(registerInstance) # for some reason, have to register instance down here!

		# SSL socket stuff
		ctx = SSL.Context(SSL.SSLv23_METHOD)
		ctx.use_privatekey_file(keyFile)
		ctx.use_certificate_file(certFile)
		self.socket = SSL.Connection(ctx, socket.socket(self.address_family, self.socket_type))
		
		old = fcntl.fcntl(self.socket.fileno(), fcntl.F_GETFD)	
		fcntl.fcntl(self.socket.fileno(), fcntl.F_SETFD, old | fcntl.FD_CLOEXEC)	
		
		self.server_bind()
		self.server_activate()

		# requests count and condition, to allow for keyboard quit via CTL-C
		self.requests = 0
		self.rCondition = Condition()


	def startup(self):
		'run until quit signaled from keyboard...'
		print '[SERVER] Starting; Press CTRL-C to quit ...'
		while True:
			try:
				self.rCondition.acquire()
				start(self.handle_request, ()) # we do this async, because handle_request blocks!
				while not self.requests:
					self.rCondition.wait(timeout=3.0)
				if self.requests: self.requests -= 1
				self.rCondition.release()
			except KeyboardInterrupt:
				print "[SERVER] Exiting..."
				return

	def get_request(self):
		request, client_address = self.socket.accept()
		self.rCondition.acquire()
		self.requests += 1
		self.rCondition.notifyAll()
		self.rCondition.release()
		return (request, client_address)

	def listMethods(self):
		'return list of method names (strings)'
		methodNames = self.funcs.keys()
		methodNames.sort()
		return methodNames

	def methodHelp(self, methodName):
		'method help'
		if methodName in self.funcs:
			return self.funcs[methodName].__doc__
		else:
			raise Exception('method "%s" is not supported' % methodName)


	def _marshaled_dispatch(self, data, dispatch_method = None, path = None, client_address=None):
		self.allow_none=True
		try:
			params, method = xmlrpclib.loads(data)
			params=(client_address,)+params

			if dispatch_method is not None:
				response = dispatch_method(method, params)
			else:

				response = self._dispatch(method, params)

			response = (response,)
			response = xmlrpclib.dumps(response, methodresponse=1,allow_none=self.allow_none, encoding=self.encoding)

		except Fault, fault:
			response = xmlrpclib.dumps(fault, allow_none=self.allow_none,encoding=self.encoding)
		except:

			exc_type, exc_value, exc_tb = sys.exc_info()
			response = xmlrpclib.dumps(xmlrpclib.Fault(1, "%s:%s" % (exc_type, exc_value)),encoding=self.encoding, allow_none=self.allow_none,)

		return response



if __name__ == '__main__':
	
	obj=imp.load_source("core","/usr/share/n4d/xmlrpc-server/core.py")
	l=obj.Core()
	server_address = ('', 9779) # (address, port)
	server = SecureDocXMLRPCServer(l, server_address, DEFAULTKEYFILE, DEFAULTCERTFILE)	
	#server.register_introspection_functions()
	'''
	for plugin in l.cm.plugins:
		for method in plugin.function:
			server.register_function(getattr(l.objects[plugin.class_name],method))
	'''
	
			
	sa = server.socket.getsockname()
	print "[SERVER] HTTPS on", sa[0], "port", sa[1]
	server.startup()


